List stake pools

Anviking opened the pull request:
# Issue Number

#1718, #1720

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Allow heterogeneous queues of LSQ queries such that we can query for stake, rewards, and pparams using the same queue.
- [x] Retrieve pool id, stake and saturation and make the API return it
- [x] Stubbed values for non-myopic member rewards, margin, cost, produced blocks

# Comments

- Actually fetching non-myopic member rewards depends on a consensus change reaching cardano-node.

- Working chain following may be a non-goal for this PR, but trying to make room for it.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Allow localStateQ to have multiple queries in the same queue

To allow both PParams, and StakeDistribution checks in the same queue.

Anviking committed:
List stake pools

- Return dummy rewards, cost, margin, productions for now
- Embed LSQ queue in cursor
- Remove mock knownPools

Anviking committed:
Remove mock pool ids from integration tests

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -433,7 +423,6 @@ codecs = clientCodecs ShelleyCodecConfig ShelleyNodeToClientVersion1
 serialisedCodecs :: MonadST m => DefaultCodecs ShelleyBlock m
 serialisedCodecs = defaultCodecs ShelleyCodecConfig ShelleyNodeToClientVersion1
 
--- | Construct a network client with the given communication channel, for the
```
Oops, mistake from rebasing.

rvl reviewed (commented):
I'm not sure if you would like a review yet, but I see that you are using the `Cursor` fields to access the `lsqQ`. Perhaps we would just need another `NetworkLayer` method for this. Because for Jormungandr, we needed to track pools, but for cardano-node, we can simply query.

Anviking committed:
Add integration test for listing stake pools

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -513,6 +523,39 @@ fromNetworkMagic :: NetworkMagic -> W.ProtocolMagic
+    Map.map (Quantity . fromIntegral)
+    . Map.mapKeys fromPoolId
+    . snd
+    . head
```
Not sure what we want to do here. We are currently only requesting rewards for one wallet at a time, so in a way perhaps we should keep it (albite with a better error than `head`)

KtorZ replied:
I agree that having `head` _here_ feels kinda wrong. If anything, the caller of `fromRewards` should make the decision about how to treat the result but this function `fromRewards` should simply do the type translation, nothing more.

Anviking replied:
b36cfe6bebe49b961356d60cd36196b47a9b901e
(But still treating this as an `error`)

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley.hs:
```diff
@@ -230,7 +233,17 @@ serveWallet
+                    randomApi
+                    icarusApi
+                    shelleyApi
+                    (\_wid -> listStakePools spl (Coin 0))
```
Ok, this could certainly be filled in now

Anviking replied:
Done in cdc4ad7

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -130,9 +130,10 @@ server
     -> ApiLayer (SeqState n IcarusKey) t IcarusKey
     -> ApiLayer (SeqState n ShelleyKey) t ShelleyKey
     -> IO [PoolId]
+    -> Server (ListStakePools ApiStakePool)
```
This is actually more of an `Handler` than a `Server`.

Anviking replied:
I'm considering using `StakePoolLayer` instead, now that the shape of it has settled 🤔

Anviking replied:
Done in cdc4ad7

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -513,6 +523,39 @@ fromNetworkMagic :: NetworkMagic -> W.ProtocolMagic
+optimumNumberOfPools = safeConvert . SL._nOpt
+  where
+    safeConvert :: Natural -> Int
+    safeConvert = fromIntegral
```
Is it though :wink: 

```
λ> safeConvert 9223372036854775808
-9223372036854775808
```

KtorZ replied:
I don't think going from an unbounded type to a bounded type will ever be "safe". Although realistic values of _optinumNumberOfPools_ are most likely within bounds.

Anviking replied:
Yes, this should be labeled "unsafeConvert" (perhaps with a comment). Not sure how I managed to get to this.

Anviking replied:
5f93fcd9cee368ea11069a0337051f8894558b45

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -549,6 +592,12 @@ fromShelleyCoin (SL.Coin c) = W.Coin $ unsafeCast c
+toShelleyCoin (W.Coin c) = SL.Coin $ safeCast c
+  where
+    safeCast :: Word64 -> Integer
+    safeCast = fromIntegral
```
This one sounds safe indeed :)

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,202 @@
+    -> Point ShelleyBlock
+    -> Coin
+    -> ExceptT ErrFetchMetrics m (Map PoolId PoolLsqMetrics)
+askNode queue pt coin = do
```
It's good to avoid _too general_ verbs like "ask" , "do", "run" etc .. in function names because they are often meaningless. `fetchMetrics` sounds a bit more transparent.

Anviking replied:
b65293b4c54fc298b10f2adcfdcde61bc5dec4e1

Went with `fetchLsqPoolMetrics`. Want to highlight that this just fetches part of the data we want with the `Lsq` part.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,202 @@
+-- | Stake Pool Data fields fetched from the node via LSQ
+data PoolLsqMetrics = PoolLsqMetrics
+    { nonMyopicMemberRewards :: Quantity "lovelace" Word64
+    , stake :: Percentage
```
perhaps: `relativeStake`

Anviking replied:
05256b73f

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,202 @@
+
+        -- Haven't figured out how to fetch non-myopic member rewards properly yet.
+        -- Let's provide a default value of 0, at least for now.
+        stakeButNoRew     = traverseMissing $ \_k s -> pure $ PoolLsqMetrics
```
NoRew? :thinking:

Anviking replied:
c6ddbea6415ea12ef38be1a43a1555a7a17e900e `Rew` -> `Rewards`.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,202 @@
+        Map.merge stakeButNoRew rewardsButNoStake bothPresent
+      where
+        -- calculate the saturation from the relative stake
+        sat s = fromRational $ (getPercentage s) / (1 / fromIntegral nOpt)
```
:+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,202 @@
+        -> Map PoolId (Quantity "lovelace" Word64)
+        -> Map PoolId PoolLsqMetrics
+    combine nOpt =
+        Map.merge stakeButNoRew rewardsButNoStake bothPresent
```
How is it actually possible with the LSQ to have cases where a poolId wouldn't be in both maps? From the excerpt above, I get that these are two separate requests, but the local state query allows for "locking" a state and making several queries on that locked state if I recall correctly? So in principle, that's sort of a guarantee that the list of pools fetched from the stake distribution is exactly the same as the list of pools get from the member rewards.

Anviking replied:
The poolId may not be in the rewards map (clarified comment in cacff5590ed5f77ba6aa0d5e6420e968733f1f99)

As for the `rewardsButNoStake` case: I _suspect_ that is impossible. (Maybe @rvl knows for sure though?)

I guess we should have some `InconcistentMetrics` similar to what we had for jormungandr, instead of the current silent `dropMissing`.

We will have more problems like this when we start folding over pool registration certificates on the chain.

Anviking replied:
> Yes, I suspect that is impossible

But I guess regardless of locking or not, we are making the queries with the same specific `Point`, so they should all be consistent if they succeed 🤔

KtorZ replied:
> I guess we should have some InconcistentMetrics similar to what we had for jormungandr, instead of the current silent dropMissing.

Not really though. In the case of Jörmungandr, we knew we could end up with inconsistent metrics because of the concurrent nature of our logic + the stateless API from Jörmungandr. So it was kind of _expected_. With cardano-node here, this is really something we do not expect, and therefore,not something we're willing to push onto the caller of this function.

Anviking committed:
Actually read wallet balance when listing pools (w/ test and swagger for failure)

Also, now passing around spl instead of a separate server for the
list stake pool endpoint.

Anviking committed:
Generalize fromRewards, rename to fromNonMyopicMemberRewards

Anviking committed:
Fix mislabeled safeConvert

Anviking committed:
Rename stake -> relativeStake

Anviking committed:
Rename stakeButNoRew -> stakeButNoRewards

Anviking committed:
askNode -> fetchLsqPoolMetrics, improve docs

Anviking committed:
Adjust comment on stakeButNoRewards merge-case

Anviking committed:
Add: STAKE_POOLS_LIST_06 - NonMyopicMemberRewards are 0 for empty wallets

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1927,7 +1927,9 @@ paths:
-      responses: *responsesListStakePools
+      responses:
+        - *responsesListStakePools
+        - *responsesErr404
```
Instead, add `responsesErr404` to the possible responses of `responsesListStakePools`:

```yaml
x-responsesListStakePools: &responsesListStakePools
  <<: *responsesErr404
  <<: *responsesErr405
  200:
    description: Ok
    content:
      application/json:
        schema:
          oneOf:
            - type: array
              items: *ApiStakePool
              title: "cardano-node stake pools"
            - type: array
              items: *ApiJormungandrStakePool
              title: "jormungandr stake pools"
```

Anviking committed:
fixup: embed responsesErr404 in responsesListStakePools itself

Anviking committed:
Move shelley LSQ code to Network module

Using a type family to allow jormungandr and haskell implementations to
be different.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -176,8 +183,13 @@ server byron icarus shelley knownPools ntp =
+            stake <- withWorkerCtx shelley wid liftE liftE $ \wrk -> do
+                (w, _, pending) <- liftHandler $ W.readWallet wrk wid
+                return $ Coin $ fromIntegral $ totalBalance pending w
+            liftHandler $ listStakePools spl stake
```
This should eventually include the reward balance as well.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -329,8 +341,36 @@ withNetworkLayer tr np addrInfo versionData action = do
+        rewardsPerAccount <- fromNonMyopicMemberRewards <$> handleQueryFailure
+            (queue `send` CmdQueryLocalState pt (OC.GetNonMyopicMemberRewards toStake))
+        pparams <- handleQueryFailure
+            (queue `send` CmdQueryLocalState pt OC.GetCurrentPParams)
```
This one feels quite redundant / awkward. It sure is needed by the caller to compute few things on the pool data, but the caller should simply call `getProtocolParameters`, don't you think ?

Anviking replied:
This code guarantees that the `pparams` are valid at the same point `pt`.

If we were to instead use
```haskell
getProtocolParameters :: m ProtocolParameters
```
We lose that control, and introduce potential race conditions. Even if unlikely, _that_ feels incredibly wrong to me.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -622,13 +658,6 @@ handleMuxError tr onResourceVanished = pure . errorType >=> \case
--- Temporary
-
-notImplemented :: HasCallStack => String -> a
-notImplemented what = error ("Not implemented: " <> what)
```
:tada:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,148 @@
+        { nonMyopicMemberRewards = Quantity 0
+        , relativeStake = s
+        , saturation = (sat s)
+        }
```
Doesn't the spec suggests something about this? Should we order pool randomly in that case?

Anviking replied:
I thought that was when we don't have enough data.

I would think that ranking according to the rewards _if_ the user had money, would be a better thing to do.

A user can't delegate without any money anyway.

We could include ROI in % and return `null` for the non-myopic member rewards.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -0,0 +1,148 @@
+    _knownPools = do
+        pt <- getTip
+        res <- runExceptT $ map fst . Map.toList
+            . combineLsqData <$> stakeDistribution nl pt dummyCoin
```
This shouldn't use the network layer but use the database. Yet, you'll need the work on my branch for that. I'll modify this function in there.

Anviking replied:
👍 Though why?

I know `knownPools` are used to validate the users's choice of delegation.

The pool worker is less likely to be in sync with the network (more downstream) than the node.

Wouldn't implementing it in terms of the LSQ make more sense?

KtorZ commented on lib/jormungandr/src/Cardano/Pool/Jormungandr/Metrics.hs:
```diff
@@ -163,7 +169,8 @@ data StakePoolLayer e m = StakePoolLayer
 -- different forks such that it's safe for readers to access it.
 monitorStakePools
-    :: Tracer IO StakePoolLog
+    :: GetStakeDistribution t IO ~ GetStakeDistribution Jormungandr IO
```
Any reason for not writing:

```hs
    :: GetStakeDistribution Jormungandr IO
```

?

Anviking replied:
Yes - allowing other targets than `Jormungandr` in the tests :S  https://github.com/input-output-hk/cardano-wallet/blob/1fab8e511360af8e567a6f2fd4e9e7b77c3df917/lib/jormungandr/test/unit/Cardano/Pool/Jormungandr/MetricsSpec.hs#L278-L281

Anviking replied:
I am thinking we should get rid of the `NetworkLayer` (or break it up) later and the need for the type families (and this weirdness)

KtorZ replied:
Okay, this is required for `module Cardano.Pool.Jormungandr.MetricsSpec` and the polymorphism on the target.

KtorZ replied:
> I am thinking we should get rid of the NetworkLayer (or break it up) later and the need for the type families (and this weirdness)

Although the source of the problem isn't much the `NetworkLayer` here, but the double-support Jormungandr / cardano-node :upside_down_face: 
Another abstraction could work better, but it'd still have to deal with backend discrepancies.
